### PR TITLE
#164829370 feature : favorite an article

### PR DESCRIPTION
**Pull request by kirega**
#### Description
This PR adds the feature that allows users to favorites an article, unfavorite an article as well as get a list of all articles they have liked.
#### Type of change
This is a feature consisting mainly of additions. It adds a favorites model, views and tests.
Below are the various endpoints and the expected payloads.
#### Note:
To test this you will need to create atleast one article as using the following payload and endpoint
##### `POST: localhost/api/articles/`
- Create an article 
```
{
    "article":
        {
            "title": "Authors Haven Test Four",
            "body": "Lorem ipsum dolor sit amet",
            "description": "Authors Haven Description Four",
            "is_published": true
        }
}
````
##### `GET` `localhost/api/articles/favorites/`
-  list all my favourite post 

```
{
    "articles": [
        {
            "id": 5,
            "article": {
                "id": 2,
                "title": "Authors Haven Test Four",
                "body": "Lorem ipsum dolor sit amet",
                "description": "Authors Haven Description Four",
                "is_published": true,
                "date_created": "2019-04-11T13:00:09.088216Z",
                "date_modified": "2019-04-11T13:00:09.088253Z",
                "slug": "authors-haven-test-four-dd3e975d4ab7",
                "read_time": "1 min",
                "author": {
                    "email": "joseph.mutiga@gmail.com",
                    "username": "kirega"
                }
            },
            "user": {
                "email": "joseph.mutiga@gmail.com",
                "username": "kirega"
            },
            "favorite": true,
            "date_created": "2019-04-11T13:03:18.830239Z",
            "date_modified": "2019-04-11T13:03:18.830270Z"
        },
        {
            "id": 1,
            "article": {
                "id": 1,
                "title": "Authors Haven Test Four",
                "body": "Lorem ipsum dolor sit amet",
                "description": "Authors Haven Description Four",
                "is_published": true,
                "date_created": "2019-04-11T07:57:10.786871Z",
                "date_modified": "2019-04-11T07:57:10.786900Z",
                "slug": "authors-haven-test-four-67f75331bfe4",
                "read_time": "1 min",
                "author": {
                    "email": "joseph.mutiga@gmail.com",
                    "username": "kirega"
                }
            },
            "user": {
                "email": "joseph.mutiga@gmail.com",
                "username": "kirega"
            },
            "favorite": true,
            "date_created": "2019-04-11T07:58:28.068952Z",
            "date_modified": "2019-04-11T07:58:28.068971Z"
        }
    ],
    "count": 2
}
```
##### `POST` `localhost/api/articles/<slug>/favorite/`
-  To favourite a post send the post request with an empty body 
- Output is as follows:
```{
    "articles": {
        "id": 1,
        "article": {
            "id": 1,
            "title": "Authors Haven Test Four",
            "body": "Lorem ipsum dolor sit amet",
            "description": "Authors Haven Description Four",
            "is_published": true,
            "date_created": "2019-04-11T07:57:10.786871Z",
            "date_modified": "2019-04-11T07:57:10.786900Z",
            "slug": "authors-haven-test-four-67f75331bfe4",
            "read_time": "1 min",
            "author": {
                "email": "joseph.mutiga@gmail.com",
                "username": "kirega"
            }
        },
        "user": {
            "email": "joseph.mutiga@gmail.com",
            "username": "kirega"
        },
        "favorite": true,
        "date_created": "2019-04-11T07:58:28.068952Z",
        "date_modified": "2019-04-11T07:58:28.068971Z"
    },
    "count": 1
}
```
##### `DELETE` `localhost/api/articles/<slug>/favorite/`
-  To unfavourite a post send the post request with an empty body 
_  Expected output is:
```
{
    "message": "Removed from favorites"
}
```
#### How Has This Been Tested?
- Clone the repo and navigate to the root directory
- Create a virtual environment by typing `virtualenv venv` and activate it
- Install requirements by typing `pip install -r dev-requirements.txt`
- Setup a database and ensure that your `.env` has the following variables:
```
export SECRET_KEY
export DEBUG
export DATABASE_URL
export SOCIAL_AUTH_GOOGLE_OAUTH2_KEY
export SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET
export SOCIAL_AUTH_TWITTER_KEY
export SOCIAL_AUTH_TWITTER_SECRET
export AH_HOST
export CLOUDINARY_NAME
export CLOUDINARY_API_KEY
export CLOUDINARY_API_SECRET
export SENDGRID_API_KEY
export FROM_EMAIL
export URL
export EMAIL_HOST=
export EMAIL_HOST_PASSWORD
export EMAIL_PORT
export EMAIL_USE_TLS
export DOMAIN

```
- Run the tests by typing `python manage.py test`

#### Checklist:
Delivers on this PT story  https://www.pivotaltracker.com/story/show/164829370

### CODE REVIEW

**Review by Raywire — DISMISSED** [review]
I like how one can be able to view the count of the favourited articles.
Screenshot 2019-04-11 at 20 32 34

File: authors/apps/articles/tests/test_favorites.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        res = self.client.post(
+            reverse('articles:favorite',
+                    kwargs={
+                        'slug': 'dfadfa'
+                    })
+        )
+        self.assertEqual(res.status_code, status.HTTP_404_NOT_FOUND)
+        # self.assertEqual(res.data['message'],
+        #                  "The article requested does not exist")
+
```
Comment (ipaullly) [inline authors/apps/articles/tests/test_favorites.py:74]: It would be lovely for our overall code quality if you could remove the commented out code

Comment (kirega) [reply authors/apps/articles/tests/test_favorites.py:74]: Oh no! I missed this comment. Thanks for pointing it out.

**Review by Kyppy — DISMISSED** [review]
Your code is neat, compact and easy to read. However, for your PR testing instructions, I think you should also include how to create an article. I had to look up Ryan's old PR just to be able to test yours. Also, you have an error in the 'post' and 'delete' URLs in your PR instructions. You typed '/favorites/' instead of '/favorite/'.

**Review by ipaullly — DISMISSED** [review]
I see all feedback was followed through, good work

**Comment by kirega** [comment]
@Kyppy The readability is the beauty of pep8. I  have noted the changes and will improve on the PR descriptions

**Review by ipaullly — DISMISSED** [review]
The readability is good, and the tests are comprehensive

**Review by Raywire — DISMISSED** [review]
Everything functions as expected.

**Review by Raywire — APPROVED** [review]
Still works great after the slug update bug fix.

**Review by MainaKamau92 — APPROVED** [review]

**Pull request closed without merging**